scipy/sparse/csgraph/_flow.pyx

```diff
@@ -670,3 +683,943 @@ cdef ITYPE_t[:] _dinic(
+    """
+    Update the potentials of the nodes in the subtree rooted at a node
+    q connected to its parent p by an edge i.
+    """
+    cdef ITYPE_t idx, d, path_len
+
+    if q == edge_targets[i]:
+        d = vertex_potentials[p] - edge_weights[i] - vertex_potentials[q]
+    else:
+        d = vertex_potentials[p] + edge_weights[i] - vertex_potentials[q]
+    path_len = _trace_subtree(q, last_descendent_dft,
+                              next_vertex_dft, subtree)
+    for idx in range(path_len):
+        q = subtree[idx]
+        vertex_potentials[q] += d
+
+cdef inline void _add_entry(
+        ITYPE_t edge_source,  # IN
+        ITYPE_t edge_target,  # IN
+        ITYPE_t flow,  # IN
+        ITYPE_t idx,  # IN
+        ITYPE_t[:] row,  # IN/OUT
+        ITYPE_t[:] col,  # IN/OUT
+        ITYPE_t[:] flow_data  # IN/OUT
+        ) nogil:
+    """
+    Adds the result values as in entry in
+    row, col and flow_data arrays.
+    """
+    row[idx] = edge_source - 1
+    col[idx] = edge_target - 1
+    flow_data[idx] = flow
+
+cdef network_simplex_result _network_simplex(
+        ITYPE_t[:] heads,  # IN
+        ITYPE_t[:] tails,  # IN
+        ITYPE_t[:] capacities,  # IN
+        ITYPE_t[:] demand,  # IN
+        ITYPE_t[:] cost,  # IN
+        ITYPE_t n_verts,  # IN
+        ITYPE_t[:] row,  # IN/OUT
+        ITYPE_t[:] col,  # IN/OUT
+        ITYPE_t[:] flow_data  # IN/OUT
+        ):
+    cdef:
+        ITYPE_t n_edges = capacities.shape[0]
+        ITYPE_t idx, faux_inf, capacities_sum, cost_sum
+        ITYPE_t j, s, t, i, tmp, p, q, n_non_zero_edges, v
+        ITYPE_t flow_cost, flow_value
+        ITYPE_t[:] edge_flow, vertex_potentials, Wne_len
+        ITYPE_t[:] parent, parent_edge, subtree_size
+        ITYPE_t[:] next_vertex_dft, prev_vertex_dft
+        ITYPE_t[:] last_descendent_dft
+        ITYPE_t[:] Wn, We, WnR, WeR, ancestors, subtree
+        return_struct r_struct
+        edge_result i_p_q
+        network_simplex_result ns_result = network_simplex_result()
+
+        ITYPE_t[:] edge_sources = np.empty(n_edges + n_verts + 1, dtype=ITYPE)
+        ITYPE_t[:] edge_targets = np.empty(n_edges + n_verts + 1, dtype=ITYPE)
+        ITYPE_t[:] edge_capacities = np.empty(n_edges + n_verts + 1,
+                                              dtype=ITYPE)
+        ITYPE_t[:] edge_weights = np.empty(n_edges + n_verts + 1, dtype=ITYPE)
+
+    edge_flow = np.zeros(n_edges + n_verts, dtype=ITYPE)
+    vertex_potentials = np.empty(n_verts + 1, dtype=ITYPE)
+    parent = np.empty(n_verts + 1, dtype=ITYPE)
+    parent_edge = np.empty(n_edges + n_verts, dtype=ITYPE)
+    subtree_size = np.empty(n_verts + 1, dtype=ITYPE)
+    next_vertex_dft = np.empty(n_verts + 1, dtype=ITYPE)
+    prev_vertex_dft = np.empty(n_verts + 1, dtype=ITYPE)
+    last_descendent_dft = np.empty(n_verts + 1, dtype=ITYPE)
+    Wn = np.empty(n_edges + 1, dtype=ITYPE)
+    We = np.empty(n_edges + 1, dtype=ITYPE)
+    Wne_len = np.empty(2, dtype=ITYPE)
```

Comment:
Would it be better to use a struct or two variables (i.e. `Wn_len`, `We_len`) instead of an array of length two here (`Wne_len`)?